### PR TITLE
feat(utils): DML generates foreign key indexes by default

### PR DESCRIPTION
**Pull request by riqwan**
**what:**

- adds default indexes on foreign keys when belongs to relationship is present

RESOLVES CORE-2404

### CODE REVIEW

File: packages/core/utils/src/dml/entity.ts [hunk 1]
```diff
@@ -122,7 +125,16 @@ export class DmlEntity<Schema extends DMLSchema> implements IDmlEntity<Schema> {
     return this
   }
 
-  indexes(indexes: EntityIndex<Schema, string | QueryCondition>[]) {
+  /**
+   * Adds indexes to be created at during model creation on the DML entity.
+   *
+   * Generated Indexes:
```
Comment (sradevski) [inline packages/core/utils/src/dml/entity.ts]:
```suggestion
   * Default Indexes:
```

File: packages/core/utils/src/dml/entity.ts [hunk 2]
```diff
@@ -122,7 +125,16 @@ export class DmlEntity<Schema extends DMLSchema> implements IDmlEntity<Schema> {
...
+   *  - A foreign key indexes will be added if any belongsTo relationships are present.
```
Comment (sradevski) [inline packages/core/utils/src/dml/entity.ts]:
```suggestion
   *  - Foreign key indexes will be added for all defined belongsTo relationships.
```

File: packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts [hunk 3]
```diff
@@ -42,3 +42,33 @@ function transformWhere(where?: string): string {
...
+*/
+export function buildForeignKeyIndexes<TSchema extends DMLSchema>(
+  schema: TSchema
+) {
+  const indexes: EntityIndex<TSchema, QueryCondition>[] = []
+
+  for (const [field, property] of Object.entries(schema)) {
+    const parsed = property.parse(field)
+    const isRelationshipType = "type" in parsed
+    const isBelongsTo = isRelationshipType && parsed.type === "belongsTo"
```
Comment (sradevski) [inline packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: Isn't the `parsed.type === "belongsTo"` sufficient?

File: packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts [hunk 4]
```diff
@@ -42,3 +42,33 @@ function transformWhere(where?: string): string {
...
+
+    if (!isBelongsTo) {
+      continue
+    }
+
+    // TODO: infer the keyname from schema
```
Comment (sradevski) [inline packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: I guess we need to address this before merging

**Review by sradevski — COMMENTED** [review]
Nice, just some minor comments. waiting for this to land for my notifications PR :)

File: packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts [hunk 5]
```diff
@@ -42,3 +42,33 @@ function transformWhere(where?: string): string {
...
+  const indexes: EntityIndex<TSchema, QueryCondition>[] = []
+
+  for (const [field, property] of Object.entries(schema)) {
+    const parsed = property.parse(field)
+    const isRelationshipType = "type" in parsed
+    const isBelongsTo = isRelationshipType && parsed.type === "belongsTo"
+
+    if (!isBelongsTo) {
+      continue
+    }
```
Comment (adrien2p) [inline packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]:
```suggestion
    if (!BelongsTo.isBelongsTo(property)) {
      continue
    }
```

Comment (adrien2p) [reply packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: I ve seen multiple places where we are checking the parsed type, but all the relation modifier have a static util to check if it is one of the relation. If some where we need to know if it is a relation of any kind we can add the same static util in the BaseRelation class like this:
```ts
const IsRelationship = Symbol.for("isRelationship")

/**
 * The BaseRelationship encapsulates the repetitive parts of defining
 * a relationship
 */
export abstract class BaseRelationship<T> implements RelationshipType<T> {
  [IsRelationship]: true = true

  static isRelationship<T>(relation: any): relation is BaseRelationship<T> {
    return !!relation?.[IsRelationship]
  }
```

**Review by adrien2p — APPROVED** [review]

File: packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts [hunk 6]
```diff
@@ -42,3 +42,33 @@ function transformWhere(where?: string): string {
 
   return where
 }
+
+/*
+  When a "belongsTo" relationship is found on the DML schema, we create a default index on the
+  foreign key name of the field.
+*/
+export function buildForeignKeyIndexes<TSchema extends DMLSchema>(
```
Comment (adrien2p) [inline packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: todoL: I believe the foreign key indexes should be managed when we transform to mikro orm entity, in this area, but I think we can prevent it from happening at the DML level where the foreign key does not yet really exists.

Comment (adrien2p) [reply packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: My thinking is that the way foreign keys are managed will depends on the adapter that will be used. In our mikro orm util, when we create the relation we also create the foreign key complement and this is where I would probably see the index being added as we discussed

File: packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts (hunk 3 above)
Comment (adrien2p) [reply packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: We have a static method on the relationship to handle this check, based on a symbole string we have on the class to be more robust

File: packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts (hunk 6 above)
Comment (riqwan) [reply packages/core/utils/src/dml/helpers/entity-builder/build-indexes.ts]: good point. Will move it to mikroorm layer.

**Pull request merged**